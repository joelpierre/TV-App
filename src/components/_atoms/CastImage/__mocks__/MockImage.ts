import { ITvShowImage } from '../../../../types/interfaces';

export const MockImage: ITvShowImage = {
  medium: "http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg",
  original: "http://static.tvmaze.com/uploads/images/original_untouched/130/325669.jpg"
};
