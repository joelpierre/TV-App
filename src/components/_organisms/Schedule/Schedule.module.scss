.Schedule {
  $grid-gap: (spacing(5) / 2);
  $grid-gap-xs: (spacing(5) / 8);
  width: 100%;

  &__container {
    background-color: transparent;
    @include max-width-container;
    padding-top: spacing(4);
    padding-bottom: spacing(2);

    @include breakpoint('sm') {
      padding-top: 0;
      padding-bottom: spacing(8);
    }
  }

  &__heading {
    font-size: 30px;
    font-family: $font-medium;
    margin-bottom: spacing(3);
    color: $color-tint-psi;
    width: 100%;
    display: flex;
    flex-wrap: wrap;

    @include breakpoint('sm') {
      align-items: flex-end;
      flex-wrap: nowrap;
      margin-bottom: spacing(6);
      color: $color-tint-alpha;
    }

    &--text-dark {
      color: $color-tint-psi;
    }
  }

  &__heading-text {
    width: 100%;
  }

  &__heading-results {
    width: auto;
    display: inline-block;
    font-size: 16px;
    margin-top: spacing();

    @include breakpoint('sm') {
      margin-top: 0;
      margin-left: auto;
    }
  }

  &__list {
    list-style-type: none;
    display: flex;
    flex-wrap: wrap;
    margin-left: -#{$grid-gap-xs};
    margin-right: -#{$grid-gap-xs};

    @include breakpoint('md') {
      margin-left: -#{$grid-gap};
      margin-right: -#{$grid-gap};
    }
  }

  &__item {
    width: percentage(1/2);
    padding: 0 $grid-gap-xs;
    margin: 0 0 spacing(2);

    @include breakpoint('sm') {
      width: percentage(1/4);
      margin: 0 0 spacing(3);
    }

    @include breakpoint('md') {
      padding: 0 $grid-gap;
      width: percentage(1/6);
    }
  }
}
