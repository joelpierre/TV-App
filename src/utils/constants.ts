export const APP_TITLE = 'Joel TVMaze App';
export const TV_MAZE_API_ROOT = 'https://api.tvmaze.com';
export const DEFAULT_COUNTRY = 'GB';
export const SITE_DESCRIPTION = 'Joel test project using the TV Maze API';
export const APP_KEYWORDS = 'TV Maze, Joel Pierre';
export const SITE_AUTHOR = 'Joel Pierre';
