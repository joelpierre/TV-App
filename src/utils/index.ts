export const isClient = (): boolean => typeof window !== undefined;

export const sanitizeEncodedChar = (text: string): string => {
  if (text && typeof text === 'string') {
    return text.replace(
      /&#(?:x([\da-f]+)|(\d+));/gi,
      (_, hex: string, dec: number): string =>
        String.fromCharCode(dec || +('0x' + hex))
    );
  }
  return String(text);
};

export const getProtocol = (): string => process.env?.NODE_ENV === 'production' ? 'https://' : 'http://';

export const getHost = (req): string => req?.headers?.host;
