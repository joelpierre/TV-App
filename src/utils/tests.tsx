// @ts-ignore
import React from 'react';

import { render, RenderOptions } from '@testing-library/react';

const AppProviders: React.FunctionComponent = ({
  children
}) => {
  return (
    <>
      {children}
    </>
  );
};

const customRender = (ui: React.ReactElement, options: Omit<RenderOptions, 'queries'> = {}) =>
  render(ui, { wrapper: AppProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
