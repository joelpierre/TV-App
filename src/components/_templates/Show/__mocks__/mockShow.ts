/* eslint-disable */
import { ITvShow } from '../../../../types/interfaces';
import { MockImage } from '../../../_atoms/CastImage/__mocks__/MockImage';

export const MockShow: ITvShow = {
  "id": 31534,
  "url": "http://www.tvmaze.com/shows/31534/the-russell-howard-hour",
  "name": "The Russell Howard Hour",
  "type": "Talk Show",
  "language": "English",
  "genres": [
    "Comedy"
  ],
  "status": "Running",
  "runtime": 60,
  "premiered": "2017-09-21",
  "officialSite": "https://www.sky.com/watch/title/series/89419c88-0079-426a-b29b-bc28bc403a58/the-russell-howard-hour",
  "schedule": {
    "time": "22:00",
    "days": [
      "Thursday"
    ]
  },
  "rating": {
    "average": 8.4
  },
  "weight": 83,
  "network": {
    "id": 63,
    "name": "Sky 1",
    "country": {
      "name": "United Kingdom",
      "code": "GB",
      "timezone": "Europe/London"
    }
  },
  "webChannel": null,
  "externals": {
    "tvrage": null,
    "thetvdb": 334902,
    "imdb": "tt7544016"
  },
  "image": MockImage,
  "summary": "<p><b>The Russell Howard Hour</b> is a weekly talk show in which Russell Howard will be chatting with celebrity guests alongside comedy stand-up. </p>",
  "updated": 1603404686,
  "_links": {
    "self": {
      "href": "http://api.tvmaze.com/shows/31534"
    },
    "previousepisode": {
      "href": "http://api.tvmaze.com/episodes/1955660"
    },
    "nextepisode": {
      "href": "http://api.tvmaze.com/episodes/1955661"
    }
  },
  "_embedded": {
    "cast": [
      {
        "person": {
          "id": 64866,
          "url": "http://www.tvmaze.com/people/64866/russell-howard",
          "name": "Russell Howard",
          "country": {
            "name": "United Kingdom",
            "code": "GB",
            "timezone": "Europe/London"
          },
          "birthday": "1980-03-23",
          "deathday": null,
          "gender": "Male",
          "image": {
            "medium": "http://static.tvmaze.com/uploads/images/medium_portrait/175/437607.jpg",
            "original": "http://static.tvmaze.com/uploads/images/original_untouched/175/437607.jpg"
          },
          "_links": {
            "self": {
              "href": "http://api.tvmaze.com/people/64866"
            }
          }
        },
        "character": {
          "id": 456592,
          "url": "http://www.tvmaze.com/characters/456592/the-russell-howard-hour-host",
          "name": "Host",
          "image": null,
          "_links": {
            "self": {
              "href": "http://api.tvmaze.com/characters/456592"
            }
          }
        },
        "self": true,
        "voice": false
      }
    ]
  }
};
