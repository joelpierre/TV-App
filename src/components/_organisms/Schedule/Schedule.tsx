import React from 'react';
import classNames from 'classnames';
import styles from './Schedule.module.scss';
import ShowCard from '@molecules/ShowCard';
import { ScheduleContext } from '../../../context/ScheduleContext';
import { getDateFromPage } from 'common/utils/index';
import usePage from '@hooks/usePage';

interface IScheduleProps {
  className?: string;
  isFirstPage?: boolean;
}

const Schedule: React.FunctionComponent<IScheduleProps> = ({
  className,
  isFirstPage = true
}) => {
  const { tvSchedule } = React.useContext(ScheduleContext);
  const { page } = usePage();
  const { date } = getDateFromPage(page);

  return (
    <article className={classNames(styles.Schedule, className)}>
      <div className={styles.Schedule__container}>
        <h2
          className={classNames(styles.Schedule__heading, {
            [styles['Schedule__heading--text-dark']]: !isFirstPage
          })}
        >
          <span className={styles['Schedule__heading-text']}>
            Shows Airing: {date}
          </span>

          <span className={styles['Schedule__heading-results']}>
            {tvSchedule.length}&nbsp;results
          </span>
        </h2>

        {tvSchedule.length > 0 && (
          <ul data-testid="schedule-list" className={styles.Schedule__list}>
            {tvSchedule.map((scheduleItem, index) => (
              <li
                key={`${scheduleItem.id} ${index}`}
                className={styles.Schedule__item}
              >
                <ShowCard
                  className={styles.Schedule__card}
                  {...scheduleItem}
                />
              </li>
            ))}
          </ul>
        )}
      </div>
    </article>
  );
};

export default Schedule;
