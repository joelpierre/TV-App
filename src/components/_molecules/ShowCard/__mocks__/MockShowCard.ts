import { ITvSchedule } from '../../../../types/interfaces';
import { MockShow } from '../../../_templates/Show/__mocks__/mockShow';
import { MockImage } from '../../../_atoms/CastImage/__mocks__/MockImage';

export const MockShowCard: ITvSchedule = {
  url: '',
  type: '',
  summary: '',
  season: 2,
  runtime: 50,
  number: 2,
  name: 'Test Show Card',
  image: MockImage.original,
  id: 123,
  airtime: '09:00',
  airstamp: 'airStamp',
  airdate: '01-01-2020',
  show: MockShow
};
