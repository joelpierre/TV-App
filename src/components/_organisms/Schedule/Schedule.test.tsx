import React from 'react';
import { render, RenderResult } from 'test-utils';

import Schedule from './Schedule';
import { MockShowCard } from '@molecules/ShowCard/__mocks__/MockShowCard';
import { APP_TITLE } from 'common/utils/constants';
import noop from 'lodash/noop';

const MockSchedule = [
  MockShowCard,
  MockShowCard,
  MockShowCard,
  MockShowCard
];

const setup = (props = {}): RenderResult => {
  const setupProps = { ...props };
  return render(<Schedule {...setupProps} />);
};

describe('<Schedule/>', () => {
  let wrapper: RenderResult;

  describe('Given TV Schedule is not and empty array', () => {
    beforeEach(() => {
      jest.spyOn(React, 'useContext').mockReset();
      jest.spyOn(React, 'useContext').mockReturnValue({
        tvSchedule: MockSchedule,
        appTitle: APP_TITLE,
        setTvSchedule: noop
      });
      wrapper = setup();
    });

    it('Should render without crashing', () => {
      expect(wrapper.container).not.toBeEmptyDOMElement();
    });

    it('should render the tvSchedule list', () => {
      expect(wrapper.getByTestId('schedule-list')).not.toBeEmptyDOMElement();
    });

    it('should match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('Given TV Schedule is empty array', () => {
    beforeEach(() => {
      jest.spyOn(React, 'useContext').mockReset();
      jest.spyOn(React, 'useContext').mockReturnValue({
        tvSchedule: [],
        appTitle: APP_TITLE,
        setTvSchedule: noop
      });
      wrapper = setup();
    });

    it('should render core component', () => {
      expect(wrapper.container).not.toBeEmptyDOMElement();
    });

    it('should not render list', () => {
      expect(wrapper.queryByTestId('schedule-list')).toBeFalsy();
    });
  });
});
