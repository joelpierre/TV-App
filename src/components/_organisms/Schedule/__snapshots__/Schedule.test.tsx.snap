// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Schedule/> Given TV Schedule is not and empty array should match snapshot 1`] = `
<DocumentFragment>
  <article
    class="Schedule"
  >
    <div
      class="Schedule__container"
    >
      <h2
        class="Schedule__heading"
      >
        <span
          class="Schedule__heading-text"
        >
          Shows Airing: Sun Oct 25 2020
        </span>
        <span
          class="Schedule__heading-results"
        >
          4Â results
        </span>
      </h2>
      <ul
        class="Schedule__list"
        data-testid="schedule-list"
      >
        <li
          class="Schedule__item"
        >
          <a
            class="Anchor ShowCard Schedule__card"
            href="/show/31534"
          >
            <article
              class="ShowCard__content"
            >
              <p
                class="ShowCard__meta"
              >
                Time: 09:00 
                <span
                  class="ShowCard__meta-time"
                >
                  (24hr)
                </span>
              </p>
              <p
                class="ShowCard__meta"
              >
                Network: Sky 1
              </p>
              <h3
                class="ShowCard__heading"
              >
                The Russell Howard Hour
              </h3>
              <div
                class="Ratings ShowCard__ratings"
              >
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star"
                />
              </div>
              <div
                class="ShowImage ShowCard__image"
              >
                <img
                  alt="The Russell Howard Hour poster"
                  class="ShowImage__image"
                  loading="lazy"
                  src="http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg"
                  width="200"
                />
              </div>
            </article>
          </a>
        </li>
        <li
          class="Schedule__item"
        >
          <a
            class="Anchor ShowCard Schedule__card"
            href="/show/31534"
          >
            <article
              class="ShowCard__content"
            >
              <p
                class="ShowCard__meta"
              >
                Time: 09:00 
                <span
                  class="ShowCard__meta-time"
                >
                  (24hr)
                </span>
              </p>
              <p
                class="ShowCard__meta"
              >
                Network: Sky 1
              </p>
              <h3
                class="ShowCard__heading"
              >
                The Russell Howard Hour
              </h3>
              <div
                class="Ratings ShowCard__ratings"
              >
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star"
                />
              </div>
              <div
                class="ShowImage ShowCard__image"
              >
                <img
                  alt="The Russell Howard Hour poster"
                  class="ShowImage__image"
                  loading="lazy"
                  src="http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg"
                  width="200"
                />
              </div>
            </article>
          </a>
        </li>
        <li
          class="Schedule__item"
        >
          <a
            class="Anchor ShowCard Schedule__card"
            href="/show/31534"
          >
            <article
              class="ShowCard__content"
            >
              <p
                class="ShowCard__meta"
              >
                Time: 09:00 
                <span
                  class="ShowCard__meta-time"
                >
                  (24hr)
                </span>
              </p>
              <p
                class="ShowCard__meta"
              >
                Network: Sky 1
              </p>
              <h3
                class="ShowCard__heading"
              >
                The Russell Howard Hour
              </h3>
              <div
                class="Ratings ShowCard__ratings"
              >
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star"
                />
              </div>
              <div
                class="ShowImage ShowCard__image"
              >
                <img
                  alt="The Russell Howard Hour poster"
                  class="ShowImage__image"
                  loading="lazy"
                  src="http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg"
                  width="200"
                />
              </div>
            </article>
          </a>
        </li>
        <li
          class="Schedule__item"
        >
          <a
            class="Anchor ShowCard Schedule__card"
            href="/show/31534"
          >
            <article
              class="ShowCard__content"
            >
              <p
                class="ShowCard__meta"
              >
                Time: 09:00 
                <span
                  class="ShowCard__meta-time"
                >
                  (24hr)
                </span>
              </p>
              <p
                class="ShowCard__meta"
              >
                Network: Sky 1
              </p>
              <h3
                class="ShowCard__heading"
              >
                The Russell Howard Hour
              </h3>
              <div
                class="Ratings ShowCard__ratings"
              >
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star Star--is-active"
                />
                <span
                  class="Star Ratings__star"
                />
              </div>
              <div
                class="ShowImage ShowCard__image"
              >
                <img
                  alt="The Russell Howard Hour poster"
                  class="ShowImage__image"
                  loading="lazy"
                  src="http://static.tvmaze.com/uploads/images/medium_portrait/130/325669.jpg"
                  width="200"
                />
              </div>
            </article>
          </a>
        </li>
      </ul>
    </div>
  </article>
</DocumentFragment>
`;
